6LoWPAN Changes and Fixes

ptcrews opened the pull request:
Hi all,


### Pull Request Overview
This commit implements several requested changes in the 6LoWPAN layer. First, the capsules were renamed from `lowpan_fragment` and `lowpan` to `sixlowpan` and `sixlowpan_compression` respectively. Additionally, several structures and traits were renamed to better reflect their real functionality. The documentation in the `lowpan` layer was further expanded, with three distinct sections corresponding to user interface, implementation, and design decisions. Finally, the multiplexing of `TxState`s was removed at this layer, and the client interface was consolidated into a single trait (`SixlowpanClient`).

### Testing Strategy

The pull request was tested using the standard `lowpan_frag_dummy` test. I saw one spurious error, which I believe was due to a userland app also using the radio.


### TODO or Help Wanted

There is still some minor functionality/questions surrounding the 6LoWPAN layer; first, whether we want to expose a `ConfigClient` interface for changing the radio. Second, how to integrate the IEEE 802.15.4 disassociation event with this layer to preserve buffers. Finally, how to better use types to avoid calls to `expect`.


### Documentation Updated

- [x] Kernel: The relevant files in `/docs` have been updated or no updates are required.
- [x] Userland: The application README has been added, updated, or no updates are required.

### Formatting

- [x] `make formatall` has been run.

hudson-ayers commented on boards/imix/src/lowpan_frag_dummy.rs:
```diff
@@ -17,13 +17,13 @@
 //! frames will prevent the test from completing successfully.
```
On the whole, this file could probably use some more comments. It is much easier to follow now than before thanks to the changed structure shortening the file and the names of the structs better reflecting their purpose, but I still think that this example implementation would benefit from comments describing certain design decisions made specific to this test. (ex: how many rx clients are initialized, what the purpose of the alarm is and how it is used, etc.)

ptcrews replied:
Alright, sure. I'll try to update this file (and the `sixlowpan_dummy` file) with additional comments.

hudson-ayers replied:
One other comment on this file: It seems like it might be useful if some of the enums like TF, SAC, and DAC were defined in the sixlowpan.rs file so that any code which wants to use the sixlowpan file doesn't have to define these enums itself, and could instead just import them.

Also: I am just making single comments as I come across things while making a UDP capsule, but if I should stop doing that and compile these into a single review I can do that - not sure what people would normally do.

ptcrews replied:
So these enums are just used internally to test the different compression methods for 6LoWPAN. The `TF` enum is the only one that might be used elsewhere (and that one would be handled by the IP layer, not the 6LoWPAN layer). The rest of these are just to test the different type of source/destination addresses we might run into/need to compress.

ptcrews commented:
First of all, thanks for the feedback. I'll try to address your comments more in depth tomorrow. I'll touch on a couple tonight, but the control flow for this layer is not the most obvious.

ptcrews commented on capsules/src/net/sixlowpan.rs:
````diff
@@ -11,91 +11,267 @@
-//! In addition to structs and their methods, this layer also defines several
-//! traits for the transmit and receive callbacks.
+//! ```
+//! Sixlowpan::transmit_packet(&self,
````
So in the case you mentioned, the `transmit_packet` function would eventually return an error, but the code for handling this case (line 488 specifically) is located elsewhere. Due to the number of edge cases in packet size processing, I tried to centralize it all (for the first frame/fragment) in the `prepare_transmit_first_fragment` function. Certainly the comment can be cleaned up however to indicate what input is expected.

ptcrews commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -11,91 +11,267 @@
-//!     information relating to the sent packet
+//! The `ip6_packet` argument contains a pointer to a buffer containing a valid
+//! IPv6 packet, while the `ip6_packet_len` argument specifies the number of
+//! bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
```
Right, so this touches on a slightly problematic part of Rust - that once you pass ownership of a mutable slice to a different layer, you cannot get the full buffer back. So in the transmit case, we need to be able to pass the full `ip6_packet` buffer to the function (and receive it back in a callback), but the packet may only take up a subsection of this buffer. So the statement `ip6_packet.len()` simply returns the full length of the buffer, but you must pass in a length that is *less* than that buffer (to prevent us indexing past the end).

So although a little confusing, this just states that the buffer you pass in *can* be longer than the length, but the length *must* be less than or equal to the total buffer length. The comment should probably be changed to make this more clear.

hudson-ayers replied:
So basically you are saying that the line "Note that 'ip6_packet.len() > ip6_packet_len'" Really should read "Note that 'ip6_packet.len() must be >= ip6_packet_len'"? Or am I still misunderstanding?

ptcrews replied:
Yeah. I can fix the comment to make this more apparent.

ptcrews commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -11,91 +11,267 @@
+// The documentation below describes the different components of the
+// fragmentation/reassembly functionality (for 6LoWPAN compression
+// documentation, please consult `capsules/src/net/sixlowpan_compression.rs`).
+//
```
I think the main difference is that the compression functionality stands better on its own as a library than fragmentation, as fragmentation (especially in the reassembly/receive case) acts like a full layer. Certainly it might be a good idea to separate out the interface into a separate file for readability however.

ptcrews commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -107,21 +283,16 @@ use kernel::hil::time;
-// Timer fire rate in seconds
-const TIMER_RATE: usize = 10;
 // Reassembly timeout in seconds
-const FRAG_TIMEOUT: usize = 60;
```
So yes, 60 seconds is a very long time. This should be either part of a consolidated `const` file containing various network constants, or be passed in as an initialization argument. The only reason it is 60 is that this is the maximum timeout defined by the 6LoWPAN spec.

ptcrews commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -11,91 +11,267 @@
+// seconds, and with a single RxState, a single lost fragment could
+// substantially hamper or delay the ability of a client to receive additional
+// packets. As a result of these two issues, the ability to add several
+// RxStates to the 6LoWPAN layer was provided. Unfortunately, this
```
I think in order to recommend a specific number we would have to do more testing/discuss it a bit, but I think 2 is probably the sweet spot for most operations (unless you are routing packets). Currently, each RxState must have a separate, 1280 byte buffer, which is huge for Tock. I think that we can certainly clean up the code by moving the RxState initialization stuff into the `sixlowpan` layer.

Also, I think that the `lowpan_frag_test` code currently only allocates a single RxState.

ptcrews commented:
Alright, cool. I think I addressed most of the comments (let me know if I misunderstood something/wasn't clear) and I'll try to address it sometime today/tomorrow. Thanks again for the feedback!

hudson-ayers commented on boards/imix/src/lowpan_frag_dummy.rs:
```diff
@@ -399,41 +371,37 @@ impl<'a, A: time::Alarm + 'a> LowpanTest<'a, A> {
                                _: &[u8],
```
I don't understand why this functions takes in this `_: &[u8]` argument which it does not appear to use at all. When `send_ipv6_packet()` is called, the `&MLP` reference is passed in here, but it still doesn't seem as though this function actually makes any use of it..

ptcrews replied:
I think this is an artifact from earlier revisions (in particular, I think that the function signature came from early revisions of the `sixlowpan_dummy.rs` file). The `MLP` buffer is for the mesh local prefix, and the test suite should probably be extended to test for different `MLP` values, which is why it was originally an argument to the function. But you're right, the function doesn't currently use it.

alevy commented:
>  I noticed that the sixlopwn_compression.rs file still uses 'unsafe' in a couple places. My understanding was that this is not allowed for capsule code?

@hudson-ayers is absolutely right about this. It bad i didn't catch this, but `unsafe` __cannot__ be in capsules. Moreover it looks like this particular _use_ of `unsafe` is incredibly dangerous (it's casting the same slice to two different references).

Instead, just parse the `IP6Header` (which I believe can be a stack variable instead of a reference). If it's a bit more expensive, or takes up a bit more space for now, so be it.

alevy commented:
Come to think of it, we're not even guaranteed the fields in IP6Header will have the correct endianess if we just transmute them. So they really _need_ to be parsed.

ptcrews commented:
So the `unsafe` call was left over from a very early implementation of 6LoWPAN compression, and that can be rewritten to follow the format defined in `net/stream.rs`. I'm not sure how much code that would be, but I think that should be in a separate pull request (since I believe it would require reformatting a large portion of the compression code). Though I definitely agree it should be done sooner rather than later.

hudson-ayers commented on boards/imix/src/lowpan_frag_dummy.rs:
```diff
@@ -399,41 +371,37 @@ impl<'a, A: time::Alarm + 'a> LowpanTest<'a, A> {
-        let tx_state = self.tx_state;
+        let frag_state = &self.frag_state;
         //frag_state.radio.config_set_pan(0xABCD);
         let ret_code = frag_state.transmit_packet(src_mac_addr,
```
Is it not an issue that calls to transmit_packet() must be placed inside an unsafe block, because it requires that IP6_DGRAM have a static lifetime? I ask because I am currently trying to write a UDP capsule which will rely on this code (as will any code that implements Thread in a capsule), but I cannot use unsafe because I am writing capsule code. Or is there a way around this that I am not aware of?

hudson-ayers replied:
So I realized this is not an issue so long as I just have the IP6_DGRAM be passed in with a static lifetime from whatever kernel code ultimately calls my UDP capsule. Sorry for unnecessary comments, I am learning as I go.

alevy replied:
Exactly. In an _actual_ driver, you'll want to just have a field of `self` that has a `&'static` in the type already, and it'll get a static variable passed in (probably by the board config).

alevy commented:
@ptcrews I'm currently taking a pass on the documentation, which is already good, but I'm just moving things around myself instead of trying to get you to guess what I'm getting at. We can revert my changes.

As I'm doing it, I'm also finding a few more issues with the high level interface, though I don't think they will require any major changes to the design. I'll comment those in a review soon.

Just wanted to keep you in the loop, I'm looking at this carefully, and I think it's close.

ptcrews commented:
Alright, thanks for the update. Let me know if you have any questions or concerns about the PR/interfaces.

alevy commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
+    /// bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
+    /// the invariant that `ip6_packet_len <= ip6_packet.len()`.
     pub fn transmit_packet(&self,
                            src_mac_addr: MacAddress,
```
Why are is the transmitter specifying the source and destination Mac addresses? Or, _how_ does the caller of `transmit_packet` know the source and destination Mac addresses? Of course 6lowpan itself needs to know them but presumably they are determined by the Mac rather than, e.g., whoever is sending a UDP packet.

ptcrews replied:
So I'm not too sure what the final design will look like, but since the concept of gateways/routers are specific to Thread, I thought it was easiest to expose this functionality to the upper layers (since they will know our current gateway and hence the mac address of the next hop). The source address could be set and controlled by the lower layers however.

alevy replied:
Doesn't the Mac layer determine what our next hop is?

I think in this case it's better to underpromise the functionality at this stage and add it as necessary, but in any case if `dst_mac_addr` should be configurable by the client of `Sixlowpan` it should be with a setter method so we can at least set it during board configuration (or elsewhere) to some default. The client of `Sixlowpan` is, e.g., the UDP driver, which has no clue what a Mac address is

ptcrews replied:
So we could go with getters/setters, but this ties into another issue we're having; how and where to expose the configuration options for each layer. Should we allow some cross-layer configuration routines (e.g. UDP calls getters/setters for 6LoWPAN layer, bypassing the IP layer)?  But in this case, the upper layers do know what the Mac address is. Regardless, the `dst_mac_addr` option definitely needs to be changeable at runtime, as an end node must be able to change their gateway.

Also, if we ever want to support routing in the future, it may make sense to expose the destination mac address as an argument to send, as we would be dynamically dispatching/sending based on Thread-level state.

alevy commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
                            ip6_packet_len: usize,
                            security: Option<(SecurityLevel, KeyId)>,
-                           tx_state: &'a TxState<'a>,
                            fragment: bool,
```
Are `security`, `fragment` and `compress` something I might want to modify on a per-packet basis? Or should they, instead, be part of the `Sixlowpan` state (and someone will modify them either in the constructor or with setter methods)? Again, how is the sender suppose to decide what options to use?

The baseline signature is `transmit_packet(ip6_packet: &'static mut [u8]) -> Result<...>` and any additional arguments need to be justified (e.g. `ip6_packet_len` is needed because of a limitation we currently have with managing static buffers).

ptcrews replied:
So security is definitely changed on a per-packet basis, and some packets cannot be fragmented (and so `fragment` was exposed as an easy way to check/fail if the packet wouldn't fit).

`compress` was exposed largely for testing reasons I think, and that can probably be removed.

alevy replied:
> So security is definitely changed on a per-packet basis

Can you be more specific? Is it just that the initial MLE messages in Thread are unencrypted, but after connection establishment everything _is_, followed by periodic key rotations? If so, this all sounds like it should be part of the `Sixlowpan` state and changed "on the side" with specific setter methods.

I'm hearing that at least both `fragment` and `compress` should be removed. If there are cases where a packet can't be fragmented and we're better off failing, it should probably be a separate method. In general, a single method with all the options exposed and one or more helper methods with some or all left out is a reasonable design.

So how is something like the `UDP` driver meant to decide which security level to use, or whether to fragment?

ptcrews replied:
So @bbbert is more familiar with the security configuration, but some packets need link-layer encryption, and some packets do not. This is handled by the Thread specification, so I think that UDP should just forward that decision up the stack, and expose it in its `send` method. I think there is also additional per-packet state that needs to be encoded in the security object, but I'm not familiar enough with how Thread deals with 802.15.4 security to say for certain.

alevy commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
-        self.tx_states.push_tail(tx_state);
-        if self.tx_busy.get() {
-            Ok(ReturnCode::SUCCESS)
+                           -> Result<ReturnCode, (ReturnCode, &'static mut [u8])> {
```
Will we ever return any variant of `Ok(...)` that is not `ReturnCode::SUCCESS`? If not, it doesn't make sense to have two kinds of return codes, and should be something more like `Result<(), (ReturnCode, &'static mut [u8]>`.

ptcrews replied:
I'll have to review the code again, but no, I don't think there's ever a variant of `Ok(...)` that is not `SUCCESS`.

alevy commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -107,21 +231,16 @@ use kernel::hil::time;
 
-pub trait ReceiveClient {
+pub trait SixlowpanClient {
     fn receive<'a>(&self, buf: &'a [u8], len: u16, result: ReturnCode);
```
It still remains to be determined, maybe, if this `'a` lifetime on the buffer makes sense, or if instead the the buffer should be managed by the client. But ok for now. We can change it later if need be

ptcrews replied:
The main advantage of this strategy is that we get to pass back an immutable buffer, and be guaranteed that after the callback completes, the RxState can take it's buffer back. More critically, I think that these buffers should *not* be managed by the clients, since dispatching on RxStates is not determined by IP address. In particular, if some client allocates multiple buffers and another client does not, the second client could still receive callbacks (and hence buffers), and simply not return them, exhausting the resources of the first client.

I'm not sure if there is a way to design the upper layers to avoid this problem, but I was thinking that the receive callbacks should all be called in sequence, requiring only a final copy from kernel space to userland.

alevy replied:
Yep, I think that's sound reasoning. Let's see if there are issues with as we build up the stack.

alevy commented on capsules/src/net/sixlowpan.rs:
```diff
@@ -166,39 +285,42 @@ fn is_fragment(packet: &[u8]) -> bool {
+    ///
+    /// # Arguments
+    ///
+    /// `tx_buf` - A buffer for storing fragments the size of a 802.15.4 frame.
```
How big does this buffer need to be? How large is a fragment size? Does it make sense for this to be encoded in the buffer type? e.g. `&'static mut [u8; 256]`

ptcrews replied:
I think that the fragment size is constant (~127 bytes, though I have to check), so it should be encoded in the buffer type.

alevy replied:
Especially since this is not going to be passed back and forth except with the 802.15.4 driver. If it ends up being a huge pain to do, no big deal, let's just document really clearly how large it needs to be

bbbert replied:
I'm not quite sure there's a point to this. Eventually you'll still need to take a static mutable slice of it (because the radio returns buffers through the callback as static mutable slices, so we cannot store them as `&'static mut [u8; BLOCK_SIZE]` internally). The most you could do is make the constructor signature require a fixed-size buffer.

bbbert replied:
Well then you would have to modify the crypto layer to only accept buffers of size 127.

alevy replied:
Oh ok, I understand. So either modifying just the constructor or at least documenting seems best.

ptcrews commented:
The latest push to this branch replaces the old parse/serialize methods for the `IP6Header` struct with the `encode`/`decode` methods, which use @bbbert 's stream abstraction. This allows it to remain consistent with the rest of the networking stack.

After the meeting(s) last week, the 6LoWPAN stack and interfaces will be substantially changed. This branch will be the base from which these changes will be implemented.

alevy commented:
OK @ptcrews, should we try to merge these changes then still? I think since we've overcome the safety issues, and now the encoding/decoding stuff is cleaner, I think it's OK to merge even though the overall structure might change.

ptcrews commented:
@alevy Just finished testing - looks like everything works, so I think we should probably try to merge the changes. Since it might be a little while before the new 6LoWPAN interface is merged, and it will build on top of this PR anyways, it might be worth it to merge.
